File: mobile/app/lib/core/data/set_logos.dart
import 'dart:convert';

import 'package:flutter/services.dart' show rootBundle;

/// Official FAB set logo URLs scraped from fabtcg.com marketing assets.
///
/// Bundled as `assets/setLogos.json` (same shape as the web app's
/// `public/setLogos.json`), keyed by TCGplayer group id — which matches
/// [CardModel.setId].
class SetLogoMap {
  const SetLogoMap(this._urlsByGroupId, [this._abbreviationsByGroupId = const {}]);

  final Map<String, String> _urlsByGroupId;
  final Map<String, String> _abbreviationsByGroupId;

  static const empty = SetLogoMap({});

  /// Parse the scraped JSON blob. Tolerates either
  /// `{ "logos": { "2724": { "logoUrl": "…" } } }` or a flat
  /// `{ "2724": "https://…" }` map.
  factory SetLogoMap.fromJson(String jsonText) {
    final decoded = jsonDecode(jsonText);
    if (decoded is! Map) return empty;

    final raw = decoded['logos'] is Map
        ? decoded['logos'] as Map
        : decoded;
    final urls = <String, String>{};
    final abbreviations = <String, String>{};
    for (final entry in raw.entries) {
      final key = entry.key.toString();
      final value = entry.value;
      String? url;
      if (value is String) {
        url = value;
      } else if (value is Map) {
        final logoUrl = value['logoUrl'];
        if (logoUrl is String) url = logoUrl;
        final abbr = value['abbreviation'];
        if (abbr is String && abbr.trim().isNotEmpty) {
          abbreviations[key] = abbr.trim();
        }
      }
      if (url != null && url.isNotEmpty) urls[key] = url;
    }
    return SetLogoMap(urls, abbreviations);
  }

  String? urlForGroupId(int? groupId) {
    if (groupId == null) return null;
    return _urlsByGroupId[groupId.toString()];
  }

  /// TCGCSV abbreviation when the bundled map has one; otherwise null.
  String? abbreviationForGroupId(int? groupId) {
    if (groupId == null) return null;
    return _abbreviationsByGroupId[groupId.toString()];
  }

  /// All distinct logo CDN URLs (for disk/memory cache warming).
  Iterable<String> get urls => _urlsByGroupId.values;

  bool get isEmpty => _urlsByGroupId.isEmpty;
  int get length => _urlsByGroupId.length;
}

/// Load the bundled set-logo map. Returns [SetLogoMap.empty] on any failure
/// so the browse UI can fall back to set names without crashing.
Future<SetLogoMap> loadSetLogoMap() async {
  try {
    final jsonText = await rootBundle.loadString('assets/setLogos.json');
    return SetLogoMap.fromJson(jsonText);
  } catch (_) {
    return SetLogoMap.empty;
  }
}

File: apps/web/src/utils/setLogos.js
// Loads official FAB set logo URLs scraped from fabtcg.com marketing assets.
// Missing entries intentionally return null so the UI can fall back to the set name.

let logosCache = null;
let logosPromise = null;

const loadSetLogosCached = async () => {
    if (logosCache) return logosCache;
    if (logosPromise) return logosPromise;
    logosPromise = (async () => {
        try {
            const response = await fetch('/setLogos.json');
            if (!response.ok) {
                logosPromise = null;
                logosCache = {};
                return logosCache;
            }
            const data = await response.json();
            const logos = data?.logos && typeof data.logos === 'object' ? data.logos : {};
            // Normalize to groupId string -> logoUrl string
            const map = {};
            for (const [groupId, entry] of Object.entries(logos)) {
                const url = typeof entry === 'string' ? entry : entry?.logoUrl;
                if (url) map[String(groupId)] = url;
            }
            logosCache = map;
            return logosCache;
        } catch (err) {
            console.warn('Failed to load set logos:', err);
            logosPromise = null;
            logosCache = {};
            return logosCache;
        }
    })();
    return logosPromise;
};

/**
 * React-friendly loader: returns { logoByGroupId, loading }.
 * logoByGroupId maps String(groupId) -> logo URL.
 */
export const loadSetLogoMap = loadSetLogosCached;

export const getSetLogoUrl = (logoByGroupId, groupId) => {
    if (!logoByGroupId || groupId == null) return null;
    return logoByGroupId[String(groupId)] || null;
};
